# NHL Puck Management — Exploratory Data Analysis and Methodology

This notebook explores the MoneyPuck skaters dataset and the NHL API 
events dataset before joining them together for modeling.

**Data Sources:**
- MoneyPuck 2025-26 season level skaters data
- NHL API play-by-play events (413,859 events across 1,312 games)

### Imports and Setup

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

# Display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:.2f}'.format)

print("Libraries loaded ✓")

Libraries loaded ✓


### Loading the Data

In [31]:
# Load MoneyPuck skaters data
skaters = pd.read_csv('../data/processed/skaters.csv')

# Load NHL API events data
events = pd.read_csv('../data/processed/events.csv')

print(f"Skaters shape: {skaters.shape}")
print(f"Events shape: {events.shape}")

Skaters shape: (4700, 154)
Events shape: (413859, 30)


In [4]:
skaters.head(10)

,playerId,season,name,team,position,situation,games_played,icetime,shifts,gameScore,onIce_xGoalsPercentage,offIce_xGoalsPercentage,onIce_corsiPercentage,offIce_corsiPercentage,onIce_fenwickPercentage,offIce_fenwickPercentage,iceTimeRank,I_F_xOnGoal,I_F_xGoals,I_F_xRebounds,I_F_xFreeze,I_F_xPlayStopped,I_F_xPlayContinuedInZone,I_F_xPlayContinuedOutsideZone,I_F_flurryAdjustedxGoals,I_F_scoreVenueAdjustedxGoals,I_F_flurryScoreVenueAdjustedxGoals,I_F_primaryAssists,I_F_secondaryAssists,I_F_shotsOnGoal,I_F_missedShots,I_F_blockedShotAttempts,I_F_shotAttempts,I_F_points,I_F_goals,I_F_rebounds,I_F_reboundGoals,I_F_freeze,I_F_playStopped,I_F_playContinuedInZone,I_F_playContinuedOutsideZone,I_F_savedShotsOnGoal,I_F_savedUnblockedShotAttempts,penalties,I_F_penalityMinutes,I_F_faceOffsWon,I_F_hits,I_F_takeaways,I_F_giveaways,I_F_lowDangerShots,I_F_mediumDangerShots,I_F_highDangerShots,I_F_lowDangerxGoals,I_F_mediumDangerxGoals,I_F_highDangerxGoals,I_F_lowDangerGoals,I_F_mediumDangerGoals,I_F_highDangerGoals,I_F_scoreAdjustedShotsAttempts,I_F_unblockedShotAttempts,I_F_scoreAdjustedUnblockedShotAttempts,I_F_dZoneGiveaways,I_F_xGoalsFromxReboundsOfShots,I_F_xGoalsFromActualReboundsOfShots,I_F_reboundxGoals,I_F_xGoals_with_earned_rebounds,I_F_xGoals_with_earned_rebounds_scoreAdjusted,I_F_xGoals_with_earned_rebounds_scoreFlurryAdjusted,I_F_shifts,I_F_oZoneShiftStarts,I_F_dZoneShiftStarts,I_F_neutralZoneShiftStarts,I_F_flyShiftStarts,I_F_oZoneShiftEnds,I_F_dZoneShiftEnds,I_F_neutralZoneShiftEnds,I_F_flyShiftEnds,faceoffsWon,faceoffsLost,timeOnBench,penalityMinutes,penalityMinutesDrawn,penaltiesDrawn,shotsBlockedByPlayer,OnIce_F_xOnGoal,OnIce_F_xGoals,OnIce_F_flurryAdjustedxGoals,OnIce_F_scoreVenueAdjustedxGoals,OnIce_F_flurryScoreVenueAdjustedxGoals,OnIce_F_shotsOnGoal,OnIce_F_missedShots,OnIce_F_blockedShotAttempts,OnIce_F_shotAttempts,OnIce_F_goals,OnIce_F_rebounds,OnIce_F_reboundGoals,OnIce_F_lowDangerShots,OnIce_F_mediumDangerShots,OnIce_F_highDangerShots,OnIce_F_lowDangerxGoals,OnIce_F_mediumDangerxGoals,OnIce_F_highDangerxGoals,OnIce_F_lowDangerGoals,OnIce_F_mediumDangerGoals,OnIce_F_highDangerGoals,OnIce_F_scoreAdjustedShotsAttempts,OnIce_F_unblockedShotAttempts,OnIce_F_scoreAdjustedUnblockedShotAttempts,OnIce_F_xGoalsFromxReboundsOfShots,OnIce_F_xGoalsFromActualReboundsOfShots,OnIce_F_reboundxGoals,OnIce_F_xGoals_with_earned_rebounds,OnIce_F_xGoals_with_earned_rebounds_scoreAdjusted,OnIce_F_xGoals_with_earned_rebounds_scoreFlurryAdjusted,OnIce_A_xOnGoal,OnIce_A_xGoals,OnIce_A_flurryAdjustedxGoals,OnIce_A_scoreVenueAdjustedxGoals,OnIce_A_flurryScoreVenueAdjustedxGoals,OnIce_A_shotsOnGoal,OnIce_A_missedShots,OnIce_A_blockedShotAttempts,OnIce_A_shotAttempts,OnIce_A_goals,OnIce_A_rebounds,OnIce_A_reboundGoals,OnIce_A_lowDangerShots,OnIce_A_mediumDangerShots,OnIce_A_highDangerShots,OnIce_A_lowDangerxGoals,OnIce_A_mediumDangerxGoals,OnIce_A_highDangerxGoals,OnIce_A_lowDangerGoals,OnIce_A_mediumDangerGoals,OnIce_A_highDangerGoals,OnIce_A_scoreAdjustedShotsAttempts,OnIce_A_unblockedShotAttempts,OnIce_A_scoreAdjustedUnblockedShotAttempts,OnIce_A_xGoalsFromxReboundsOfShots,OnIce_A_xGoalsFromActualReboundsOfShots,OnIce_A_reboundxGoals,OnIce_A_xGoals_with_earned_rebounds,OnIce_A_xGoals_with_earned_rebounds_scoreAdjusted,OnIce_A_xGoals_with_earned_rebounds_scoreFlurryAdjusted,OffIce_F_xGoals,OffIce_A_xGoals,OffIce_F_shotAttempts,OffIce_A_shotAttempts,xGoalsForAfterShifts,xGoalsAgainstAfterShifts,corsiForAfterShifts,corsiAgainstAfterShifts,fenwickForAfterShifts,fenwickAgainstAfterShifts
0,8478043,2025,Sam Lafferty,CHI,C,other,29,85.00,1.00,-0.75,0.00,0.88,0.00,0.67,0.00,0.85,29.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,1.00,0.00,0.00,0.00,1.00,0.00,0.00,0.00,1.00,0.00,0.00,471.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.0

In [6]:
events.head(10)

,game_id,season,game_date,home_team,away_team,home_score_final,away_score_final,event_id,event_type,sort_order,period,period_type,time_in_period,time_remaining,situation_code,home_team_defending_side,x_coord,y_coord,zone_code,player_id,event_owner_team_id,hitting_player_id,hittee_player_id,shooting_player_id,scoring_player_id,away_score,home_score,player_name,position_code,player_team_id
0,2025020912,20252026,2026-02-25,TBL,TOR,4,2,9,period-start,10,1,REG,00:00,20:00,1551,left,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2025020912,20252026,2026-02-25,TBL,TOR,4,2,8,faceoff,11,1,REG,00:00,20:00,1551,left,0.00,0.00,N,NaN,10.00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2025020912,20252026,2026-02-25,TBL,TOR,4,2,61,giveaway,12,1,REG,00:18,19:42,1551,left,21.00,-29.00,N,8477503.00,10.00,NaN,NaN,NaN,NaN,NaN,NaN,Max Domi,C,10.00
3,2025020912,20252026,2026-02-25,TBL,TOR,4,2,53,missed-shot,13,1,REG,00:22,19:38,1551,left,62.00,21.00,O,NaN,14.00,NaN,NaN,8476826.00,NaN,NaN,NaN,NaN,NaN,NaN
4,2025020912,20252026,2026-02-25,TBL,TOR,4,2,64,giveaway,14,1,REG,00:27,19:33,1551,left,73.00,-37.00,O,8476826.00,14.00,NaN,NaN,NaN,NaN,NaN,NaN,Yanni Gourde,C,14.00
5,2025020912,20252026,2026-02-25,TBL,TOR,4,2,65,takeaway,16,1,REG,00:31,19:29,1551,left,63.00,19.00,O,8476878.00,14.00,NaN,NaN,NaN,NaN,NaN,NaN,Zemgus Girgensons,C,14.00
6,2025020912,20252026,2026-02-25,TBL,TOR,4,2,67,giveaway,18,1,REG,00:32,19:28,1551,left,91.00,25.00,O,8476878.00,14.00,NaN,NaN,NaN,NaN,NaN,NaN,Zemgus Girgensons,C,14.00
7,2025020912,20252026,2026-02-25,TBL,TOR,4,2,62,missed-shot,25,1,REG,00:51,19:09,1551,left,80.00,-13.00,O,NaN,14.00,NaN,NaN,8478010.00,NaN,NaN,NaN,NaN,NaN,NaN
8,2025020912,20252026,2026-02-25,TBL,TOR,4,2,66,shot-on-goal,26,1,REG,01:08,18:52,1551,left,-78.00,-4.00,O,NaN,10.00,NaN,NaN,8481711.00,NaN,NaN,NaN,NaN,NaN,NaN
9,2025020912,20252026,2026-02-25,TBL,TOR,4,2,81,takeaway,36,1,REG,01:47,18:13,1551,left,-35.00,-36.00,D,8477404.00,14.00,NaN,NaN,NaN,NaN,NaN,NaN,Jake Guentzel,C,14.00


## MoneyPuck Skaters Data Exploration

First we explore the MoneyPuck skaters dataset to understand the distribution 
of puck management metrics before any modeling.

In [10]:
print(f"Shape: {skaters.shape}")
print(f"\nUnique players: {skaters['name'].nunique()}")
print(f"\nGames played range: {skaters['games_played'].min()} to {skaters['games_played'].max()}")

Shape: (4700, 154)

Unique players: 939

Games played range: 1 to 83


83 games means it's likely a player was traded mid-season!

### Filtering our skaters dataset to only show 5v5 stats

In [13]:
skaters = skaters[skaters['situation'] == '5on5'].copy()
skaters.head(10)

,playerId,season,name,team,position,situation,games_played,icetime,shifts,gameScore,onIce_xGoalsPercentage,offIce_xGoalsPercentage,onIce_corsiPercentage,offIce_corsiPercentage,onIce_fenwickPercentage,offIce_fenwickPercentage,iceTimeRank,I_F_xOnGoal,I_F_xGoals,I_F_xRebounds,I_F_xFreeze,I_F_xPlayStopped,I_F_xPlayContinuedInZone,I_F_xPlayContinuedOutsideZone,I_F_flurryAdjustedxGoals,I_F_scoreVenueAdjustedxGoals,I_F_flurryScoreVenueAdjustedxGoals,I_F_primaryAssists,I_F_secondaryAssists,I_F_shotsOnGoal,I_F_missedShots,I_F_blockedShotAttempts,I_F_shotAttempts,I_F_points,I_F_goals,I_F_rebounds,I_F_reboundGoals,I_F_freeze,I_F_playStopped,I_F_playContinuedInZone,I_F_playContinuedOutsideZone,I_F_savedShotsOnGoal,I_F_savedUnblockedShotAttempts,penalties,I_F_penalityMinutes,I_F_faceOffsWon,I_F_hits,I_F_takeaways,I_F_giveaways,I_F_lowDangerShots,I_F_mediumDangerShots,I_F_highDangerShots,I_F_lowDangerxGoals,I_F_mediumDangerxGoals,I_F_highDangerxGoals,I_F_lowDangerGoals,I_F_mediumDangerGoals,I_F_highDangerGoals,I_F_scoreAdjustedShotsAttempts,I_F_unblockedShotAttempts,I_F_scoreAdjustedUnblockedShotAttempts,I_F_dZoneGiveaways,I_F_xGoalsFromxReboundsOfShots,I_F_xGoalsFromActualReboundsOfShots,I_F_reboundxGoals,I_F_xGoals_with_earned_rebounds,I_F_xGoals_with_earned_rebounds_scoreAdjusted,I_F_xGoals_with_earned_rebounds_scoreFlurryAdjusted,I_F_shifts,I_F_oZoneShiftStarts,I_F_dZoneShiftStarts,I_F_neutralZoneShiftStarts,I_F_flyShiftStarts,I_F_oZoneShiftEnds,I_F_dZoneShiftEnds,I_F_neutralZoneShiftEnds,I_F_flyShiftEnds,faceoffsWon,faceoffsLost,timeOnBench,penalityMinutes,penalityMinutesDrawn,penaltiesDrawn,shotsBlockedByPlayer,OnIce_F_xOnGoal,OnIce_F_xGoals,OnIce_F_flurryAdjustedxGoals,OnIce_F_scoreVenueAdjustedxGoals,OnIce_F_flurryScoreVenueAdjustedxGoals,OnIce_F_shotsOnGoal,OnIce_F_missedShots,OnIce_F_blockedShotAttempts,OnIce_F_shotAttempts,OnIce_F_goals,OnIce_F_rebounds,OnIce_F_reboundGoals,OnIce_F_lowDangerShots,OnIce_F_mediumDangerShots,OnIce_F_highDangerShots,OnIce_F_lowDangerxGoals,OnIce_F_mediumDangerxGoals,OnIce_F_highDangerxGoals,OnIce_F_lowDangerGoals,OnIce_F_mediumDangerGoals,OnIce_F_highDangerGoals,OnIce_F_scoreAdjustedShotsAttempts,OnIce_F_unblockedShotAttempts,OnIce_F_scoreAdjustedUnblockedShotAttempts,OnIce_F_xGoalsFromxReboundsOfShots,OnIce_F_xGoalsFromActualReboundsOfShots,OnIce_F_reboundxGoals,OnIce_F_xGoals_with_earned_rebounds,OnIce_F_xGoals_with_earned_rebounds_scoreAdjusted,OnIce_F_xGoals_with_earned_rebounds_scoreFlurryAdjusted,OnIce_A_xOnGoal,OnIce_A_xGoals,OnIce_A_flurryAdjustedxGoals,OnIce_A_scoreVenueAdjustedxGoals,OnIce_A_flurryScoreVenueAdjustedxGoals,OnIce_A_shotsOnGoal,OnIce_A_missedShots,OnIce_A_blockedShotAttempts,OnIce_A_shotAttempts,OnIce_A_goals,OnIce_A_rebounds,OnIce_A_reboundGoals,OnIce_A_lowDangerShots,OnIce_A_mediumDangerShots,OnIce_A_highDangerShots,OnIce_A_lowDangerxGoals,OnIce_A_mediumDangerxGoals,OnIce_A_highDangerxGoals,OnIce_A_lowDangerGoals,OnIce_A_mediumDangerGoals,OnIce_A_highDangerGoals,OnIce_A_scoreAdjustedShotsAttempts,OnIce_A_unblockedShotAttempts,OnIce_A_scoreAdjustedUnblockedShotAttempts,OnIce_A_xGoalsFromxReboundsOfShots,OnIce_A_xGoalsFromActualReboundsOfShots,OnIce_A_reboundxGoals,OnIce_A_xGoals_with_earned_rebounds,OnIce_A_xGoals_with_earned_rebounds_scoreAdjusted,OnIce_A_xGoals_with_earned_rebounds_scoreFlurryAdjusted,OffIce_F_xGoals,OffIce_A_xGoals,OffIce_F_shotAttempts,OffIce_A_shotAttempts,xGoalsForAfterShifts,xGoalsAgainstAfterShifts,corsiForAfterShifts,corsiAgainstAfterShifts,fenwickForAfterShifts,fenwickAgainstAfterShifts
2,8478043,2025,Sam Lafferty,CHI,C,5on5,29,14856.00,351.00,0.83,0.40,0.42,0.47,0.45,0.45,0.45,318.00,10.20,0.85,0.84,2.63,0.42,5.31,3.95,0.80,0.83,0.78,0.00,1.00,9.00,5.00,6.00,20.00,2.00,1.00,2.00,0.00,0.00,0.00,4.00,7.00,8.00,13.00,2.00,4.00,25.00,35.00,2.00,15.00,10.00,3.00,1.00,0.38,0.27,0.20,1.00,0.00,0.00,19.70,14.00,13.89,3.00,0.18,0.09,0.38,0.64,0.63,0.62,351.00,34.00,18.00,54.00,245.00,49.00,58.00,35.00,209.00,25.00,26.00,72051.00,4.00,2.00,1.00,5.00,85.54,5.

In [14]:
# Distribution of icetime
print(skaters['icetime'].describe())
print(f"\n25th percentile: {skaters['icetime'].quantile(0.25):.0f} seconds")
print(f"Median: {skaters['icetime'].quantile(0.5):.0f} seconds")

count     940.00
mean    41217.30
std     27454.06
min       312.00
25%     14525.25
50%     44722.50
75%     62997.25
max     98762.00
Name: icetime, dtype: float64

25th percentile: 14525 seconds
Median: 44722 seconds


### Ice Time Distribution & Minimum Threshold

Ice time is measured in seconds. Key percentiles:
- 25th percentile: ~242 minutes of 5on5 ice time
- Median: ~745 minutes
- Max: ~1,646 minutes

**Confidence threshold: 242 minutes (25th percentile)**

Players below this threshold will not be excluded but instead have their 
score pulled toward the league average using confidence weighting:

`final_score = (confidence × player_score) + ((1 - confidence) × league_average)`

Where confidence scales from 0 to 1 based on ice time relative to the threshold.
This means fringe players who barely played aren't penalised unfairly but also 
don't skew the rankings based on a tiny sample size.

#### What columns are relevant to our inquiry? 

In [18]:
print(skaters.columns.tolist())


['playerId', 'season', 'name', 'team', 'position', 'situation', 'games_played', 'icetime', 'shifts', 'gameScore', 'onIce_xGoalsPercentage', 'offIce_xGoalsPercentage', 'onIce_corsiPercentage', 'offIce_corsiPercentage', 'onIce_fenwickPercentage', 'offIce_fenwickPercentage', 'iceTimeRank', 'I_F_xOnGoal', 'I_F_xGoals', 'I_F_xRebounds', 'I_F_xFreeze', 'I_F_xPlayStopped', 'I_F_xPlayContinuedInZone', 'I_F_xPlayContinuedOutsideZone', 'I_F_flurryAdjustedxGoals', 'I_F_scoreVenueAdjustedxGoals', 'I_F_flurryScoreVenueAdjustedxGoals', 'I_F_primaryAssists', 'I_F_secondaryAssists', 'I_F_shotsOnGoal', 'I_F_missedShots', 'I_F_blockedShotAttempts', 'I_F_shotAttempts', 'I_F_points', 'I_F_goals', 'I_F_rebounds', 'I_F_reboundGoals', 'I_F_freeze', 'I_F_playStopped', 'I_F_playContinuedInZone', 'I_F_playContinuedOutsideZone', 'I_F_savedShotsOnGoal', 'I_F_savedUnblockedShotAttempts', 'penalties', 'I_F_penalityMinutes', 'I_F_faceOffsWon', 'I_F_hits', 'I_F_takeaways', 'I_F_giveaways', 'I_F_lowDangerShots', 'I_F_

**Key Puck Management Columns:**
- `I_F_giveaways` — total giveaways at 5on5
- `I_F_takeaways` — total takeaways at 5on5
- `I_F_dZoneGiveaways` — defensive zone giveaways
- `I_F_dZoneShiftStarts` — defensive zone shift starts

In [21]:
cols = ['name', 'icetime', 'I_F_giveaways', 'I_F_dZoneGiveaways', 
        'I_F_takeaways', 'I_F_dZoneShiftStarts']

print(skaters[cols].describe())
print(f"\nMissing values:\n{skaters[cols].isnull().sum()}")

       icetime  I_F_giveaways  I_F_dZoneGiveaways  I_F_takeaways  \
count   940.00         940.00              940.00         940.00   
mean  41217.30          36.49               16.78          10.65   
std   27454.06          27.48               17.08           8.78   
min     312.00           0.00                0.00           0.00   
25%   14525.25          11.00                4.00           3.00   
50%   44722.50          35.00               11.00           9.00   
75%   62997.25          56.00               22.00          17.00   
max   98762.00         129.00               82.00          44.00   

       I_F_dZoneShiftStarts  
count                940.00  
mean                  92.79  
std                   68.09  
min                    0.00  
25%                   27.00  
50%                   94.00  
75%                  142.00  
max                  310.00  

Missing values:
name                    0
icetime                 0
I_F_giveaways           0
I_F_dZoneGiveaways    

## NHL API Events Data Exploration

Exploring the play-by-play events dataset built from all 1,312 regular 
season games. This dataset will be used for the giveaway above expected model.

In [32]:
# Quick overview of events data
print(f"Shape: {events.shape}")
print(f"\nEvent types:\n{events['event_type'].value_counts()}")
print(f"\nDate range: {events['game_date'].min()} to {events['game_date'].max()}")
print(f"\nUnique games: {events['game_id'].nunique()}")

Shape: (413859, 30)

Event types:
event_type
faceoff                73496
shot-on-goal           65370
stoppage               57202
hit                    53658
blocked-shot           40866
giveaway               40366
missed-shot            39168
takeaway               12099
penalty                 9692
goal                    8350
period-start            4381
period-end              4381
delayed-penalty         3384
game-end                1312
shootout-complete        119
failed-shot-attempt       15
Name: count, dtype: int64

Date range: 2025-10-07 to 2026-04-16

Unique games: 1312


In [23]:
events = events.sort_values(['game_date', 'game_id', 'sort_order']).reset_index(drop=True)
events.head(10)

,game_id,season,game_date,home_team,away_team,home_score_final,away_score_final,event_id,event_type,sort_order,period,period_type,time_in_period,time_remaining,situation_code,home_team_defending_side,x_coord,y_coord,zone_code,player_id,event_owner_team_id,hitting_player_id,hittee_player_id,shooting_player_id,scoring_player_id,away_score,home_score,player_name,position_code,player_team_id
0,2025020001,20252026,2025-10-07,FLA,CHI,3,2,52,period-start,8,1,REG,00:00,20:00,1551,right,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2025020001,20252026,2025-10-07,FLA,CHI,3,2,51,faceoff,11,1,REG,00:00,20:00,1551,right,0.00,0.00,N,NaN,16.00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2025020001,20252026,2025-10-07,FLA,CHI,3,2,101,blocked-shot,12,1,REG,00:22,19:38,1551,right,-61.00,3.00,D,NaN,13.00,NaN,NaN,8473419.00,NaN,NaN,NaN,NaN,NaN,NaN
3,2025020001,20252026,2025-10-07,FLA,CHI,3,2,8,stoppage,13,1,REG,00:34,19:26,1551,right,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2025020001,20252026,2025-10-07,FLA,CHI,3,2,55,faceoff,16,1,REG,00:34,19:26,1551,right,-69.00,22.00,D,NaN,16.00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
5,2025020001,20252026,2025-10-07,FLA,CHI,3,2,74,giveaway,18,1,REG,00:46,19:14,1551,right,-41.00,38.00,O,8482113.00,13.00,NaN,NaN,NaN,NaN,NaN,NaN,Anton Lundell,C,13.00
6,2025020001,20252026,2025-10-07,FLA,CHI,3,2,851,giveaway,24,1,REG,01:10,18:50,1551,right,-13.00,-31.00,N,8484783.00,16.00,NaN,NaN,NaN,NaN,NaN,NaN,Artyom Levshunov,D,16.00
7,2025020001,20252026,2025-10-07,FLA,CHI,3,2,67,blocked-shot,25,1,REG,01:13,18:47,1551,right,-64.00,-16.00,D,NaN,13.00,NaN,NaN,8482713.00,NaN,NaN,NaN,NaN,NaN,NaN
8,2025020001,20252026,2025-10-07,FLA,CHI,3,2,9,stoppage,26,1,REG,01:15,18:45,1551,right,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
9,2025020001,20252026,2025-10-07,FLA,CHI,3,2,68,faceoff,28,1,REG,01:15,18:45,1551,right,-69.00,-22.00,O,NaN,13.00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### Giveaway Analysis

Exploring the 40,366 giveaway events to understand location, 
context and what typically follows a turnover.

In [33]:
# Filter to giveaways only
giveaways = events[events['event_type'] == 'giveaway'].copy()

print(f"Total giveaways: {len(giveaways)}")
print(f"\nGiveaways by zone:\n{giveaways['zone_code'].value_counts()}")
print(f"\nGiveaways by situation:\n{giveaways['situation_code'].value_counts()}")

Total giveaways: 40366

Giveaways by zone:
zone_code
D    18643
O    13836
N     7887
Name: count, dtype: int64

Giveaways by situation:
situation_code
1551    35422
1451     1923
1541     1835
651       318
1560      308
1441      282
1331      174
1460       23
641        16
1531       15
1351       15
1431       10
1341        8
551         6
1450        4
541         4
1550        3
Name: count, dtype: int64


In [34]:
giveaways = events[events['event_type'] == 'giveaway']
giveaways = giveaways[giveaways['situation_code'] == 1551].copy()
print(f"Total events: {len(events)}")
print(f"5on5 giveaways: {len(giveaways)}")

Total events: 413859
5on5 giveaways: 35422


We only want to keep 5v5 giveaways!

### What Happens After a Giveaway?

We want to understand what typically follows a giveaway. We can do this by looking at the next event within the same game after each giveaway event.

In [35]:
# Create an empty list to store what happens after each giveaway
next_event_types = []

# Loop through every giveaway
for idx, row in giveaways.iterrows():
    
    # Find all events in the same game that happened after this giveaway
    after = events[
        (events['game_id'] == row['game_id']) & 
        (events['sort_order'] > row['sort_order'])
    ]
    
    # Grab the very next event
    if len(after) > 0:
        next_event_types.append(after.iloc[0]['event_type'])
    else:
        next_event_types.append('none')
        
pd.Series(next_event_types).value_counts()

shot-on-goal       7398
hit                7143
giveaway           5416
blocked-shot       4622
missed-shot        4239
stoppage           3136
takeaway           1563
goal                790
penalty             461
delayed-penalty     367
period-end          285
faceoff               2
Name: count, dtype: int64

### What Follows a Giveaway?

After analysing all 35,422 5on5 giveaways, the immediate next event breakdown is:

| Next Event | Count | % |
|---|---|---|
| Shot on goal | 7,398 | 21% |
| Hit | 7,143 | 20% |
| Giveaway | 5,416 | 15% |
| Blocked shot | 4,622 | 13% |
| Missed shot | 4,239 | 12% |
| Stoppage | 3,136 | 9% |
| Takeaway | 1,563 | 4% |
| Goal | 790 | 2% |
| Other | 1,289 | 4% |

**Key finding:** Shots (shot on goal, blocked shot, missed shot, goal) 
immediately follow a giveaway 48% of the time.

## Methodology & Metrics

### Overview
This project builds a composite puck management score for every NHL skater 
in the 2025-26 regular season. The goal is to move beyond simple counting 
statistics and build a context-adjusted, data-driven framework that fairly 
evaluates how well players manage the puck.

All analysis is restricted to 5on5 situations (situation code 1551) to ensure 
a level playing field since special teams situations distort puck management metrics 
since players are under fundamentally different levels of pressure.

---

### Data Sources

**MoneyPuck Season Level Skaters Data**
Season level aggregated statistics for all 939 NHL skaters in the 2025-26 
regular season. Filtered to 5on5 situations only. Provides clean, reliable 
counting statistics for giveaways, takeaways, zone starts and ice time.

**NHL API Play-by-Play Data**
Raw event level data pulled directly from the NHL's public API for all 1,312 
regular season games (413,859 total events). Each giveaway event includes exact 
x/y coordinates, zone, manpower situation, period, and timing. 

---

### Metrics

We settled on three distinct, non-redundant metrics that each measure a 
genuinely different dimension of puck management:

**Metric 1 — Giveaway Rate per 60**

How often a player turns the puck over per 60 minutes of 5on5 ice time. 
Lower is better. We initially considered a net turnover metric 
(takeaway rate - giveaway rate) but found that raw giveaway and takeaway 
counts correlate at r=0.83, meaning players involved in more puck battles 
tend to have both higher giveaways and takeaways. Including both separately 
would double count the same behaviour.

**Metric 2 — Giveaway Above Expected**

Our EDA showed that shots immediately follow 48% of all giveaways. However 
not all giveaways are equally dangerous — a turnover in the offensive zone 
corner is fundamentally different from one in front of your own net. 

For each giveaway we look at the next shot within 10 seconds and record its 
xGoal value (the probability it becomes a goal). We will then train a regression 
model using the giveaway's context — x/y coordinates, zone, period, time 
remaining, and score differential — to learn what the expected danger is for 
a giveaway in any given situation. 

A player's giveaway above expected score is the difference between the actual 
danger their giveaways caused and what was expected given the situations they 
were in. A negative score means their giveaways caused less danger than 
expected — good. A positive score means they turned it over in more dangerous 
spots than average — bad.

This directly addresses the key weakness of manual weighting systems where 
arbitrary values are assigned to different event types without data backing.

**Metric 3 — Defensive Zone Giveaway Rate (adjusted for zone starts)**

While giveaway above expected captures how dangerous turnovers are, it doesn't 
explicitly isolate how frequently a player turns it over in their own end. We 
calculate dzone giveaways per defensive zone shift start rather than per 60 
minutes. This adjusts for zone deployment — a player who starts 70% of shifts 
in the defensive zone will naturally face more opportunities to give it away 
there. Dividing by dzone shift starts makes this a fair comparison across all 
player types.

---

### Composite Score

Each metric is independently normalised using MAD (Median Absolute Deviation) 
z-scoring within position groups as forwards and defensemen are scored separately 
since they have fundamentally different roles and usage patterns. The MAD z-score forumla
is shown below: 

$$\text{MAD Z-Score} = 0.6745 \times \frac{x - \text{median}(x)}{\text{MAD}(x)}$$

A confidence weight is then applied based on each player's 5on5 ice time. 
Players below the 25th percentile threshold (242 minutes) have their score 
pulled toward the positional average rather than being excluded entirely:

The three normalised, confidence weighted metrics are then combined into a 
single composite puck management score. Equal weights are applied initially 
and adjusted based on correlation analysis and model performance.